R/derive_advs_params.R

```diff
@@ -0,0 +1,600 @@
+        filter(PARAMCD == param) %>%
+        pull(`_unit`) %>%
+        unique()
+
+      if (length(units_found) != 1L) {
+        cli_abort(
+          "Multiple units {.val {units_found}} found for {.val {param}}.
+          Please review and update the units."
+        )
+      }
+
+      param_units[[param]] <- units_found
+    }
+
+    # If the input parameters are measured in different units
+    # but are convertible from one to another (and this kind of conversion supported)
+    # then modify the formula in order to perform units conversion on the fly
+
+    if (param_units[[denominator_code]] != param_units[[numerator_code]]) {
+      # Find conversion factor for denominator
+      conv_factor <- get_conv_factor(
+        param_units[[denominator_code]],
+        param_units[[numerator_code]]
+      )
+
+      if (!is.na(conv_factor)) {
+        ratio_formula <- expr(
+          !!sym(paste0("AVAL.", numerator_code)) / (
+            !!sym(paste0("AVAL.", denominator_code)) * !!conv_factor
+          )
+        )
+
+        cli_alert_info(
+          "Unit conversion performed for {.val {denominator_code}}. Values converted from
+          {.val {param_units[[denominator_code]]}} to {.val {param_units[[numerator_code]]}}.",
+          wrap = TRUE
+        )
+      }
+    }
+  }
+
+  ### Identify constant parameters ----
+
+  parameters <- c(numerator_code, denominator_code)
+  constant_parameters <- NULL
+
+  if (constant_numerator) {
+    constant_parameters <- c(constant_parameters, numerator_code)
+
+    parameters <- parameters %>%
+      .[!. == numerator_code]
+  }
+
+  if (constant_denominator) {
+    constant_parameters <- c(constant_parameters, denominator_code)
+
+    parameters <- parameters %>%
+      .[!. == denominator_code] %>%
+      ifelse(length(.) == 0, NULL, .)
+  }
+
+  ### Call the core {admiral} function to derive Ratio parameter ----
+
+  derive_param_computed(
+    dataset,
+    filter = !!filter,
+    parameters = parameters,
+    by_vars = by_vars,
+    set_values_to = exprs(
+      AVAL = !!ratio_formula,
+      !!!set_values_to
+    ),
+    constant_parameters = constant_parameters,
+    constant_by_vars = constant_by_vars
+  )
+}
+
+#' Unit conversion
+#'
+#' @name unit-conversion
+#' @keywords internal
+NULL
+#> NULL
+
+#' @description `get_conv_factor()` extracts a conversion factor for a pair of units.
+#' Returns `NA` if units are not supported/convertible.
+#'
+#' @rdname unit-conversion
+#' @keywords internal
+get_conv_factor <- function(from_unit, to_unit) {
+  # Get all conversion factors supported
+  conv_factors_all <- get_conv_factors_all()
+
+  # Return conversion factor if units are supported and convertible
+  for (unit_category in names(conv_factors_all)) {
+    if (all(c(from_unit, to_unit) %in% names(conv_factors_all[[unit_category]]))) {
+      return(
+        conv_factors_all[[unit_category]][[from_unit]] /
+          conv_factors_all[[unit_category]][[to_unit]]
+      )
+    }
+  }
```

Comment:
I think we should keep it silent for other potential ratios where input parameters are not supposed to be convertible, like Height (cm) to Weight (kg) or so. For `derive_param_waisthip` and `derive_param_waisthgt` this is no longer a problem since I've updated them to assert units using `assert_unit()` with only those, which are supported.